File: backend/src/models/mailModel.js
const { PARCEL_STATUS } = require('../config/constants');

class Mail {
  constructor(
    mailId,
    userId,
    senderName,
    receiverName,
    receiverAddress,
    weight,
    status = PARCEL_STATUS.PENDING,
    createdBy
  ) {
    this.mailId = mailId;
    this.userId = userId;
    this.senderName = senderName;
    this.receiverName = receiverName;
    this.receiverAddress = receiverAddress;
    this.weight = Number(weight);
    this.status = status;
    this.createdBy = createdBy; // admin who created the parcel
    this.createdAt = new Date().toISOString();
    this.updatedAt = new Date().toISOString();
  }

  static fromFirestore(doc) {
    const data = doc.data();
    const mail = new Mail(
      doc.id,
      data.userId,
      data.senderName,
      data.receiverName,
      data.receiverAddress,
      data.weight,
      data.status,
      data.createdBy
    );
    mail.createdAt = data.createdAt;
    mail.updatedAt = data.updatedAt;
    return mail;
  }

  toFirestore() {
    return {
      mailId: this.mailId,
      userId: this.userId,
      senderName: this.senderName,
      receiverName: this.receiverName,
      receiverAddress: this.receiverAddress,
      weight: this.weight,
      status: this.status,
      createdBy: this.createdBy,
      createdAt: this.createdAt,
      updatedAt: this.updatedAt,
    };
  }
}

module.exports = Mail;

File: Frontend/post_app/lib/models/mail_model.dart
import 'package:post_app/models/enums.dart';

class MailModel {
  final String mailId;
  final String userId;
  final String senderName;
  final String receiverName;
  final String receiverAddress;
  final double weight;
  final ParcelStatus status;
  final String createdBy;
  final DateTime createdAt;
  final DateTime updatedAt;

  MailModel({
    required this.mailId,
    required this.userId,
    required this.senderName,
    required this.receiverName,
    required this.receiverAddress,
    required this.weight,
    required this.status,
    required this.createdBy,
    required this.createdAt,
    required this.updatedAt,
  });

  factory MailModel.fromJson(Map<String, dynamic> json) {
    return MailModel(
      mailId: json['mailId'] as String,
      userId: json['userId'] as String,
      senderName: json['senderName'] as String,
      receiverName: json['receiverName'] as String,
      receiverAddress: json['receiverAddress'] as String,
      weight: (json['weight'] as num).toDouble(),
      status: ParcelStatus.fromJson(json['status'] as String),
      createdBy: json['createdBy'] as String,
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt: DateTime.parse(json['updatedAt'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'mailId': mailId,
      'userId': userId,
      'senderName': senderName,
      'receiverName': receiverName,
      'receiverAddress': receiverAddress,
      'weight': weight,
      'status': status.toJson(),
      'createdBy': createdBy,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }
}

// --- Request/Response Models for Mail ---

class CreateMailRequest {
  final String userId;
  final String senderName;
  final String receiverName;
  final String receiverAddress;
  final double weight;
  final String receiverEmail;

  CreateMailRequest({
    required this.userId,
    required this.senderName,
    required this.receiverName,
    required this.receiverAddress,
    required this.weight,
    required this.receiverEmail,
  });

  Map<String, dynamic> toJson() {
    return {
      'userId': userId,
      'senderName': senderName,
      'receiverName': receiverName,
      'receiverAddress': receiverAddress,
      'weight': weight,
    };
  }
}

class AdminUpdateMailStatusRequest {
  final ParcelStatus status;

  AdminUpdateMailStatusRequest({required this.status});

  Map<String, dynamic> toJson() {
    return {
      'status': status.toJson(),
    };
  }
}

class CreateMailResponse {
  final String message;
  final MailModel mail;

  CreateMailResponse({
    required this.message,
    required this.mail,
  });

  factory CreateMailResponse.fromJson(Map<String, dynamic> json) {
    return CreateMailResponse(
      message: json['message'] as String,
      mail: MailModel.fromJson(json['mail'] as Map<String, dynamic>),
    );
  }
}

class SendTrackingEmailRequest {
  final String recipientEmail;
  final String trackingNumber;
  final String receiverName;
  final String senderName;

  SendTrackingEmailRequest({
    required this.recipientEmail,
    required this.trackingNumber,
    required this.receiverName,
    required this.senderName,
  });

  Map<String, dynamic> toJson() {
    return {
      'recipientEmail': recipientEmail,
      'trackingNumber': trackingNumber,
      'receiverName': receiverName,
      'senderName': senderName,
    };
  }
}